include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+        Iter               choicesEnd)
+        : Knob(flagName), mChoices(choicesBegin, choicesEnd)
+    {
+        PPX_ASSERT_MSG(defaultIndex < mChoices.size(), "defaultIndex is out of range");
+
+        SetDefaultAndIndex(defaultIndex);
+    }
+
+    template <typename Container>
+    KnobDropdown(
+        const std::string& flagName,
+        size_t             defaultIndex,
+        const Container&   container)
+        : KnobDropdown(flagName, defaultIndex, std::begin(container), std::end(container)) {}
+
+    void Draw() override
+    {
+        if (ImGui::BeginCombo(GetDisplayName().c_str(), mChoices.at(mIndex).c_str())) {
+            for (size_t i = 0; i < mChoices.size(); ++i) {
+                bool isSelected = (i == mIndex);
+                if (ImGui::Selectable(mChoices.at(i).c_str(), isSelected)) {
+                    if (i != mIndex) { // A new choice is selected
+                        mIndex = i;
+                    }
+                }
+                if (isSelected) {
+                    ImGui::SetItemDefaultFocus();
+                }
+            }
+            ImGui::EndCombo();
+        }
+    }
+
+    void ResetToDefault() override { SetIndex(mDefaultIndex); }
+
+    std::string GetFlagHelpText() const override
+    {
+        std::string choiceStr = "";
+        for (const auto& choice : mChoices) {
+            choiceStr += '\"' + choice + '\"' + "|";
+        }
+        if (!choiceStr.empty()) {
+            choiceStr.pop_back();
+        }
+
+        std::string flagHelpText = "--" + GetFlagName() + " <" + choiceStr + ">";
+        if (!GetFlagHelp().empty()) {
+            flagHelpText += " : " + GetFlagHelp();
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Expected commandline flag format:
+    // --flag_name <str>
+    void UpdateFromFlags(const CliOptions& opts) override
+    {
+        SetDefaultAndIndex(opts.GetExtraOptionValueOrDefault(GetFlagName(), GetValue()));
+    }
+
+    size_t   GetIndex() const { return mIndex; }
+    const T& GetValue() const { return mChoices[mIndex]; }
+
+    // Used for when mIndex needs to be updated outside of UI
+    void SetIndex(size_t newI)
+    {
+        if (!IsValidIndex(newI)) {
+            PPX_LOG_ERROR(GetFlagName() << " does not have this index in allowed choices: " << newI);
+            return;
+        }
+
+        mIndex = newI;
+    }
+
+    // Used for setting from flags
+    void SetIndex(const std::string& newVal)
+    {
+        auto temp = std::find(mChoices.cbegin(), mChoices.cend(), newVal);
+        if (temp == mChoices.cend()) {
+            PPX_LOG_ERROR(GetFlagName() << " does not have this value in allowed range: " << newVal);
+            return;
+        }
+
+        mIndex = std::distance(mChoices.cbegin(), temp);
+    }
+
+private:
+    bool IsValidIndex(size_t i)
+    {
+        if (i >= mChoices.size()) {
+            return false;
+        }
+        return true;
+    }
+
+    void SetDefaultAndIndex(size_t newI)
```

Comment:
nit: in general google c++ style discourage abbreviations except for simple counters in loops. so `newIndex`, `index`, `newValue`, etc.
But that's really a nit of a nit 😊